src/bson_util/async_decode.rs

```diff
@@ -0,0 +1,214 @@
+            Some(Binary) => {
+                let len = read_i32(reader).await?;
+                if len < 0 || len > MAX_BSON_SIZE {
+                    return Err(DecoderError::InvalidLength(
+                        len as usize,
+                        format!("Invalid binary length of {}", len),
+                    ));
+                }
+                let subtype = BinarySubtype::from(read_u8(reader).await?);
+                let mut data = Vec::with_capacity(len as usize);
+                reader.take(len as u64).read_to_end(&mut data).await?;
+                Ok(Bson::Binary(subtype, data))
+            }
+            Some(ObjectId) => {
+                let mut objid = [0; 12];
+                for x in &mut objid {
+                    *x = read_u8(reader).await?;
+                }
+                Ok(Bson::ObjectId(oid::ObjectId::with_bytes(objid)))
+            }
+            Some(Boolean) => Ok(Bson::Boolean(read_u8(reader).await? != 0)),
+            Some(NullValue) => Ok(Bson::Null),
+            Some(RegularExpression) => {
+                let pat = read_cstring(reader).await?;
+                let opt = read_cstring(reader).await?;
+                Ok(Bson::RegExp(pat, opt))
+            }
+            Some(JavaScriptCode) => read_string(reader, utf8_lossy)
+                .await
+                .map(Bson::JavaScriptCode),
+            Some(JavaScriptCodeWithScope) => {
+                // disregard the length:
+                //     using Read::take causes infinite type recursion
+                read_i32(reader).await?;
+
+                let code = read_string(reader, utf8_lossy).await?;
+                let scope = decode_document(reader).await?;
+                Ok(Bson::JavaScriptCodeWithScope(code, scope))
+            }
+            Some(Integer32Bit) => read_i32(reader).await.map(Bson::I32),
+            Some(Integer64Bit) => read_i64(reader).await.map(Bson::I64),
+            Some(TimeStamp) => read_i64(reader).await.map(Bson::TimeStamp),
+            Some(UtcDatetime) => {
+                // The int64 is UTC milliseconds since the Unix epoch.
+                let time = read_i64(reader).await?;
+
+                let sec = time / 1000;
+                let tmp_msec = time % 1000;
+                let msec = if tmp_msec < 0 {
+                    1000 - tmp_msec
+                } else {
+                    tmp_msec
+                };
+
+                match Utc.timestamp_opt(sec, (msec as u32) * 1_000_000) {
+                    LocalResult::None => Err(DecoderError::InvalidTimestamp(time)),
+                    LocalResult::Ambiguous(..) => Err(DecoderError::AmbiguousTimestamp(time)),
+                    LocalResult::Single(t) => Ok(Bson::UtcDatetime(t)),
+                }
+            }
+            Some(Symbol) => read_string(reader, utf8_lossy).await.map(Bson::Symbol),
+            Some(Undefined) | Some(DbPointer) | Some(MaxKey) | Some(MinKey) | None => {
```

Comment:
being unable to read these seems like a problem that we will want to fix before 1.0. As with decimal128, we need to be able to at least read/round trip all the BSON types.

Reply:
I don't think reading them is necessarily required for 1.0, as adding in the support later isn't a breaking change, but given how easy it would be to implement each of these four, I'm not opposed to doing it now. Would you prefer that I do it as part of this ticket or a follow-up?

Reply:
A follow up would be better I think.

Also, from [Bernie's comments on the roadmap](https://docs.google.com/document/d/1A48UMVSpd-l3QTQs4Alh7g5KZZ2XX9zOdqzRViOHSh4/edit?disco=AAAADiol9nc), we were actually supposed to implement [RUST-217](https://jira.mongodb.org/browse/RUST-217) for the alpha so that users could understand why decimal128s (and other types, I guess) couldn't be parsed while we hadn't implemented them. The implication with that seemed to be that once 1.0 came out, we would support decoding all those types. Either way, that's not work we need to do in this PR, but it is something to keep in mind as we get closer to 1.0.

Reply:
Filed [RUST-311](https://jira.mongodb.org/browse/RUST-311) for the follow on work.